### PR TITLE
Fix E1102 / ``not-callable`` false positive for property that returns a lambda function conditionally

### LINKED ISSUE #5931: False positive: `not-callable` for property that returns a lambda function conditionally
### Bug description

```python
class C:
    def __init__(self):
        self._x = None

    @property
    def foo(self):
        if self._x is None:
            self._x = lambda: None
        return self._x

c = C()
c.foo()
```

### Configuration

_No response_

### Command used

```shell
pylint myfile.py | grep E1102
```

### Pylint output

```shell
myfile.py:12:0: E1102: c.foo is not callable (not-callable)
```

### Expected behavior

No error. The code runs when run under python.

### Pylint version

```shell
pylint 2.12.2
astroid 2.9.3
Python 3.10.1 (main, Dec  6 2021, 22:25:40) [Clang 13.0.0 (clang-1300.0.29.3)]
```

### OS / Environment

MacOS Monterey

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by tushar-deepsource**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Update E1102 to only be raised when no inferred value is callable.

Resolves #5931

### CODE REVIEW

**Comment by tushar-deepsource** [comment]
https://github.com/PyCQA/pylint/pull/6068/files#diff-9b76776694be02a09ca25a56adee0b3087ad5b86dfafd318923c4bf008c131c3R1233-R1238

This is the part that isn't covered, but it's hard to come up with a case for this, that won't break in the future.

File: pylint/checkers/typecheck.py [hunk 1]
```diff
@@ -1227,20 +1227,25 @@ def _check_uninferable_call(self, node):
...
             if decorated_with_property(attr):
-
                 try:
-                    all_returns_are_callable = all(
-                        return_node.callable() or return_node is astroid.Uninferable
-                        for return_node in attr.infer_call_result(node)
-                    )
-                except astroid.InferenceError:
-                    continue
+                    call_results = list(attr.infer_call_result(node))
```
Comment (jacobtylerwalls) [inline pylint/checkers/typecheck.py:1231]: Can you place the except directly after this so that the try only wraps what might fail?

**Comment by srijan-paul** [comment]
zamn :^)

File: pylint/checkers/typecheck.py [hunk 2]
```diff
@@ -1227,20 +1227,22 @@ def _check_uninferable_call(self, node):
...
                     continue
 
-                if not all_returns_are_callable:
-                    self.add_message(
-                        "not-callable", node=node, args=node.func.as_string()
-                    )
-                    break
+                if all(
+                    return_node is astroid.Uninferable for return_node in call_results
+                ):
```
Comment (jacobtylerwalls) [inline pylint/checkers/typecheck.py:1237]:
```suggestion
                if astroid.Uninferable in call_results:
```

File: pylint/checkers/typecheck.py [hunk 3]
```diff
@@ -1227,20 +1227,22 @@ def _check_uninferable_call(self, node):
...
+                    # We were unable to infer return values of the call, skipping
```
Comment (jacobtylerwalls) [inline pylint/checkers/typecheck.py:1238]:
```suggestion
                    # We were unable to infer some return value of the call, skipping
```

**Review by jacobtylerwalls — COMMENTED** [review]
nice! one suggestion to loosen this even more.

**Comment by jacobtylerwalls** [comment]
Thanks for looking into it. Darn.

I still think long-term we could consider making that change, though. There's no restriction on what you can set with the property. You can set something that's callable, and then `not-callable` becomes sort of wrong. But no one is asking for that change right now :-)

**Comment by DanielNoord** [comment]
@tushar-deepsource Coming into this pretty late but I was taking a look at stale PRs and found #5593. Could you quickly check if this solves that issue?

**Comment by tushar-deepsource** [comment]
@DanielNoord This change only affects properties, so no it doesn't solve that issue.

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Only needs a changelog entry!

Didn't see this was only for `properties`. But that makes sense :)

**Comment by tushar-deepsource** [comment]
@DanielNoord hopefully this is correct?

File: ChangeLog [hunk 4]
```diff
@@ -9,6 +9,10 @@ Release date: TBA
 ..
   Put new features here and also in 'doc/whatsnew/2.14.rst'
 
+* Only raise ``not-callable`` when all the inferred values of an expression are not callable.
```
Comment (DanielNoord) [inline ChangeLog:12]:
```suggestion
* Only raise ``not-callable`` when all the inferred values of a property are not callable.
```

File: doc/whatsnew/2.14.rst [hunk 5]
```diff
@@ -38,6 +38,10 @@ Extensions
 Other Changes
 =============
 
+* Only raise ``not-callable`` when all the inferred values of an expression are not callable.
```
Comment (DanielNoord) [inline doc/whatsnew/2.14.rst:41]:
```suggestion
* Only raise ``not-callable`` when all the inferred values of a property are not callable.
```

**Review by DanielNoord — APPROVED** [review]
Almost 😄 

But thanks @tushar-deepsource! Wonderful fix!

**Comment by jacobtylerwalls** [comment]
Are we moving the changelog entry to 2.13.5?

**Pull request merged**